README.md
# sleep-parser
[![crates.io version][1]][2] [![build status][3]][4]
[![downloads][5]][6] [![docs.rs docs][7]][8]

Parse [Dat protocol SLEEP
files](https://github.com/datproject/docs/blob/master/papers/sleep.md).

- [Documentation][8]
- [Crates.io][2]

## Usage
```rust
extern crate sleep_parser as sleep_parser;

use sleep_parser::{FileType, HashType, Header};
use std::fs::File;
use std::io::{BufRead, BufReader};

let file = File::open("tests/fixtures/content.bitfield").unwrap();
let mut reader = BufReader::with_capacity(32, file);
let buffer = reader.fill_buf().unwrap();
let header = Header::from_vec(&buffer).unwrap();
assert!(header.is_bitfield());
```

## Format
```txt,ignore
<32 byte header>
  <4 byte magic string: 0x05025702>
  <1 byte version number: 0>
  <2 byte entry size: 40>
  <1 byte algorithm name length prefix: 7>
  <7 byte algorithm name: BLAKE2b>
  <17 zeroes>
<40 byte entries>
  <32 byte BLAKE2b hash>
  <8 byte Uint64BE children leaf byte length>
```

## Installation
```sh
$ cargo add sleep-parser
```

## License
[MIT](./LICENSE-MIT) or [Apache-2.0](./LICENSE-APACHE)

[1]: https://img.shields.io/crates/v/sleep-parser.svg?style=flat-square
[2]: https://crates.io/crates/sleep-parser
[3]: https://img.shields.io/travis/datrs/sleep-parser.svg?style=flat-square
[4]: https://travis-ci.org/datrs/sleep-parser
[5]: https://img.shields.io/crates/d/sleep-parser/master.svg?style=flat-square
[6]: https://crates.io/crates/sleep-parser
[7]: https://img.shields.io/badge/docs-latest-blue.svg?style=flat-square
[8]: https://docs.rs/sleep-parser


CHANGELOG.md
## 2018-12-28, Version 0.8.0
### Commits
- [[`f2fd8ba77a`](https://github.com/datrs/sleep-parser/commit/f2fd8ba77a73d1b6d3860f91f632013785c6a41b)] (cargo-release) version 0.8.0 (Yoshua Wuyts)
- [[`5927b5e69b`](https://github.com/datrs/sleep-parser/commit/5927b5e69b52a620a2162b507d3b11cf1c67ae22)] remove trailing zero verification (#13) (Yoshua Wuyts)
- [[`bb5cd2c9ba`](https://github.com/datrs/sleep-parser/commit/bb5cd2c9bada1f1f539ea5a8eb3fef2cb36e1a60)] inline parsing (#12) (Yoshua Wuyts)
- [[`4544c57897`](https://github.com/datrs/sleep-parser/commit/4544c5789795125a0ec7dcf995798e2b4dc89d53)] edition 2018 (#11) (Yoshua Wuyts)
- [[`fae3dece83`](https://github.com/datrs/sleep-parser/commit/fae3dece83818e1ccadcc083c09a27246b8449f9)] Update changelog (Yoshua Wuyts)

### Stats
```diff
 CHANGELOG.md    | 24 ++++++++++++++++++++++++
 Cargo.toml      |  3 ++-
 src/lib.rs      | 37 ++++++++++++++-----------------------
 tests/header.rs |  8 ++++----
 4 files changed, 44 insertions(+), 28 deletions(-)
```


## 2018-10-18, Version 0.7.0
### Commits
- [[`563b87b65f`](https://github.com/datrs/sleep-parser/commit/563b87b65f018f8ef5b70e8a18d03fe6e4663348)] (cargo-release) version 0.7.0 (Yoshua Wuyts)
- [[`603749a204`](https://github.com/datrs/sleep-parser/commit/603749a204c88033c96824727d585837069ba958)] work on stable (#10) (Yoshua Wuyts)
- [[`2c80b79a27`](https://github.com/datrs/sleep-parser/commit/2c80b79a27a5f2a556f83b7a1925669272ecba05)] Keep up with modern times in clippy invocation (#9) (Szabolcs Berecz)
- [[`db1d66d294`](https://github.com/datrs/sleep-parser/commit/db1d66d2941b7316a5faa9ce8134e4e4dd008c01)] Update .github (Yoshua Wuyts)
- [[`d23967a196`](https://github.com/datrs/sleep-parser/commit/d23967a196cfeb2bc41bdc000e58354cda2e5e2d)] ignore .dat (Yoshua Wuyts)
- [[`e9858cd35a`](https://github.com/datrs/sleep-parser/commit/e9858cd35a234b632c69511a56266ad0c53a22f2)] (cargo-release) start next development iteration 0.6.1-alpha.0 (Yoshua Wuyts)

### Stats
```diff
 .github/ISSUE_TEMPLATE.md                 | 40 +++-----------------------------
 .github/ISSUE_TEMPLATE/bug_report.md      | 23 ++++++++++++++++++-
 .github/ISSUE_TEMPLATE/feature_request.md | 30 ++++++++++++++++++++++++-
 .github/ISSUE_TEMPLATE/question.md        | 18 ++++++++++++++-
 .gitignore                                |  1 +-
 .travis.yml                               | 24 ++++++++++---------
 Cargo.toml                                |  2 +-
 README.md                                 |  4 +--
 src/lib.rs                                | 19 ++++++++++++---
 9 files changed, 108 insertions(+), 53 deletions(-)
```




tests/fns.rs
extern crate sleep_parser;

use sleep_parser::*;

#[test]
fn bitfield() {
  assert!(create_bitfield().is_bitfield());
}

#[test]
fn signatures() {
  assert!(create_signatures().is_signatures());
}

#[test]
fn tree() {
  assert!(create_tree().is_tree());
}


tests/header.rs
extern crate sleep_parser;

use sleep_parser::*;
use std::fs::File;
use std::io::Read;

#[test]
fn new_header() {
  Header::new(FileType::Tree, 40, HashType::BLAKE2b);
}

fn read_header_bytes(file_name: &str) -> Result<[u8; 32], std::io::Error> {
  let mut file = File::open(file_name)?;
  let mut buffer = [0u8; 32];
  file.read_exact(&mut buffer).map(|_| buffer)
}

#[test]
fn from_vec_content_bitfield() {
  let buffer = read_header_bytes("tests/fixtures/content.bitfield").unwrap();
  let header = Header::from_vec(&buffer).unwrap();
  assert!(header.is_bitfield());
  assert_eq!(header.to_vec(), buffer);
}

#[test]
fn from_vec_content_signatures() {
  let buffer = read_header_bytes("tests/fixtures/content.signatures").unwrap();
  let header = Header::from_vec(&buffer).unwrap();
  assert!(header.is_signatures());
  assert_eq!(header.to_vec(), buffer);
}

#[test]
fn from_vec_content_tree() {
  let buffer = read_header_bytes("tests/fixtures/content.tree").unwrap();
  let header = Header::from_vec(&buffer).unwrap();
  assert!(header.is_tree());
  assert_eq!(header.to_vec(), buffer);
}

#[test]
fn from_vec_metadata_bitfield() {
  let buffer = read_header_bytes("tests/fixtures/metadata.bitfield").unwrap();
  let header = Header::from_vec(&buffer).unwrap();
  assert!(header.is_bitfield());
  assert_eq!(header.to_vec(), buffer);
}

#[test]
fn from_vec_metadata_signatures() {
  let buffer = read_header_bytes("tests/fixtures/metadata.signatures").unwrap();
  let header = Header::from_vec(&buffer).unwrap();
  assert!(header.is_signatures());
  assert_eq!(header.to_vec(), buffer);
}

#[test]
fn from_vec_metadata_tree() {
  let buffer = read_header_bytes("tests/fixtures/metadata.tree").unwrap();
  let header = Header::from_vec(&buffer).unwrap();
  assert!(header.is_tree());
  assert_eq!(header.to_vec(), buffer);
}

#[test]
fn to_vec() {
  let header = Header::new(FileType::Tree, 40, HashType::BLAKE2b);
  assert_eq!(
    header.to_vec(),
    vec![
      5, 2, 87, 2, 0, 0, 40, 7, 66, 76, 65, 75, 69, 50, 98, 0, 0, 0, 0, 0, 0,
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    ]
  );
}

#[test]
fn issue_3() {
  // https://github.com/datrs/sleep-parser/issues/3

  let data = b"\x05\x02W\x01\x00\xb0\xb0\xb0\xb0\xb0\xb0\xb0\xb0\xb0\xb0\xb0\xfb\x03p\xb0\xb0\xb0\xb0\xb0\xb0\xb0\xb0\xbb9\xb0\xf5\xf5";
  assert!(Header::from_vec(data).is_err());

  let data = b"\x05\x02W\x01\x00\x00\x00\x12\x12\x12\x00\x00S\xc3\xcf\x8a2\xcc\xd1\xce9\xc4K\x9343\x00602\xb5\x07";
  assert!(Header::from_vec(data).is_err());
}

#[test]
fn invalid_algorithm() {
  fn mk_header(prefix: &[u8]) -> [u8; 32] {
    let mut h = [0u8; 32];
    h[0..prefix.len()].clone_from_slice(prefix);
    h
  }

  assert!(
    Header::from_vec(&mk_header(b"\x05\x02W\x01\x00\x00\x28\x01B")).is_err()
  );
  assert!(
    Header::from_vec(&mk_header(b"\x05\x02W\x01\x00\x00\x28\x01B")).is_err()
  );
  assert!(Header::from_vec(
    b"\x05\x02W\x01\x00\x00\x28\x19BLAKE2bXXXXXXXXXXXXXXXXXX"
  )
  .is_err());
}


src/header.rs
extern crate byteorder;

use self::byteorder::{BigEndian, ReadBytesExt, WriteBytesExt};
use failure::Error;
use std::io::Cursor;

/// Algorithm used for hashing the data.
#[derive(Debug, PartialEq)]
pub enum HashType {
  /// [BLAKE2b](https://blake2.net/) hashing algorithm.
  BLAKE2b,
  /// [Ed25519](https://ed25519.cr.yp.to/) hashing algorithm.
  Ed25519,
  /// No hashing used.
  None,
}

/// Type of file.
///
/// `signatures`, `bitfield` and `tree` are the three SLEEP files. There are two
///additional files, `key`, and `data`, which do not contain SLEEP file headers
///and store plain serialized data for easy access. `key` stores the public key
///that is described by the `signatures` file, and `data` stores the raw chunk
///data that the `tree` file contains the hashes and metadata.
#[derive(Debug, PartialEq)]
pub enum FileType {
  /// The bitfield describes which pieces of data you have, and which nodes in
  /// the tree file have been written.  This file exists as an index of the tree
  /// and data to quickly figure out which pieces of data you have or are
  /// missing. This file can be regenerated if you delete it, so it is
  /// considered a materialized index.
  BitField,
  /// A SLEEP formatted 32 byte header with data entries being 64 byte
  /// signatures.
  Signatures,
  /// A SLEEP formatted 32 byte header with data entries representing a
  /// serialized Merkle tree based on the data in the data storage layer. All
  /// the fixed size nodes written in in-order tree notation. The header
  /// algorithm string for `tree` files is `BLAKE2b`. The entry size is 40
  /// bytes.
  Tree,
}

/// SLEEP Protocol version.
#[derive(Debug)]
pub enum ProtocolVersion {
  /// The version specified as per the paper released in 2017-09.
  V0,
}

/// Structural representation of 32 byte SLEEP headers.
#[derive(Debug)]
pub struct Header {
  /// Type of file.
  pub file_type: FileType,
  /// Version of the SLEEP protocol.
  pub protocol_version: ProtocolVersion,
  /// Size of each piece of data in the file body.
  pub entry_size: u16,
  /// Algorithm used for hashing the content.
  pub hash_type: HashType,
}

const HEADER_LENGTH: usize = 32;
const MAX_ALGORITHM_NAME_LENGTH: usize = HEADER_LENGTH - 8;

/// According to https://github.com/datproject/docs/blob/master/papers/sleep.md trailing bytes
/// should be zeros, so garbage is probably fine too.
const VERIFY_TRAILING_ZEROS: bool = false;

impl Header {
  /// Create a new `Header`.
  pub fn new(
    file_type: FileType,
    entry_size: u16,
    hash_type: HashType,
  ) -> Self {
    Header {
      file_type,
      entry_size,
      hash_type,
      protocol_version: ProtocolVersion::V0,
    }
  }

  /// Parse a 32 byte buffer slice into a valid Header.
  pub fn from_vec(buffer: &[u8]) -> Result<Header, Error> {
    ensure!(buffer.len() == 32, "buffer should be 32 bytes");

    let mut rdr = Cursor::new(buffer);
    let byte = rdr.read_u8().unwrap();
    ensure!(
      byte == 5,
      "The first byte of a SLEEP header should be '5', found {}",
      byte
    );

    let byte = rdr.read_u8().unwrap();
    ensure!(
      byte == 2,
      "The second byte of a SLEEP header should be '2', found {}",
      byte
    );

    let byte = rdr.read_u8().unwrap();
    ensure!(
      byte == 87,
      "The third byte of a SLEEP header should be '87', found {}",
      byte
    );

    let file_type = match rdr.read_u8().unwrap() {
      0 => FileType::BitField,
      1 => FileType::Signatures,
      2 => FileType::Tree,
      num => bail!(
        "The fourth byte '{}' does not belong to any known SLEEP file type",
        num
      ),
    };

    let protocol_version = match rdr.read_u8().unwrap() {
      0 => ProtocolVersion::V0,
      num => bail!(
        "The fifth byte '{}' does not belong to any known SLEEP protocol protocol_version",
        num
      ),
    };

    // Read entry size which will inform how many bytes to read next.
    let entry_size = rdr.read_u16::<BigEndian>().unwrap();

    // Read out the "entry_size" bytes into a string.
    // NOTE(yw): there should be a more concise way of doing this.
    let hash_name_len = rdr.read_u8().unwrap() as usize;
    let current = rdr.position() as usize;

    ensure!(
      hash_name_len <= MAX_ALGORITHM_NAME_LENGTH,
      "Algorithm name is too long: {} (max: {})",
      hash_name_len,
      MAX_ALGORITHM_NAME_LENGTH
    );

    let hash_name_upper = current + hash_name_len;
    ensure!(
      buffer.len() >= hash_name_upper,
      "Broken parser: algorithm name is out of bounds: {} {}",
      hash_name_upper,
      buffer.len()
    );

    let buf_slice = &buffer[current..hash_name_upper];
    rdr.set_position(hash_name_upper as u64 + 1);
    let algo = ::std::str::from_utf8(buf_slice).map_err(|e| {
      format_err!("The algorithm string was invalid utf8 encoded: {:?}", e)
    })?;

    let hash_type = match algo {
      "BLAKE2b" => HashType::BLAKE2b,
      "Ed25519" => HashType::Ed25519,
      "" => HashType::None,
      name => bail!("Unexpected algorithm name: {}", name),
    };

    if VERIFY_TRAILING_ZEROS {
      for index in rdr.position()..32 {
        let byte = rdr.read_u8().unwrap();
        ensure!(
          byte == 0,
          "The remainder of the header should be zero-filled. Found byte '{}' at position '{}'.",
          byte, index
        );
      }
    }

    Ok(Header {
      protocol_version,
      entry_size,
      file_type,
      hash_type,
    })
  }

  /// Convert a `Header` into a `Vec<u8>`. Use this to persist a header back to
  /// disk.
  pub fn to_vec(&self) -> Vec<u8> {
    let mut wtr = Vec::with_capacity(32);

    wtr.extend_from_slice(&[5u8, 2, 87]);

    let file_type = match self.file_type {
      FileType::BitField => 0,
      FileType::Signatures => 1,
      FileType::Tree => 2,
    };
    wtr.write_u8(file_type).unwrap();

    let protocol_version = match self.protocol_version {
      ProtocolVersion::V0 => 0,
    };
    wtr.write_u8(protocol_version).unwrap();

    wtr.write_u16::<BigEndian>(self.entry_size).unwrap();

    let hash_type = match self.hash_type {
      HashType::BLAKE2b => "BLAKE2b",
      HashType::Ed25519 => "Ed25519",
      HashType::None => "",
    };
    let hash_type = hash_type.as_bytes();
    wtr.write_u8(hash_type.len() as u8).unwrap();
    wtr.extend_from_slice(hash_type);

    for _ in wtr.len()..wtr.capacity() {
      wtr.write_u8(0).unwrap();
    }
    wtr
  }

  /// Check whether the header is formatted as a `.bitfield`.
  pub fn is_bitfield(&self) -> bool {
    self.entry_size == 3328
      && self.file_type == FileType::BitField
      && self.hash_type == HashType::None
  }

  /// Check whether the header is formatted as a `.signatures`.
  pub fn is_signatures(&self) -> bool {
    self.entry_size == 64
      && self.file_type == FileType::Signatures
      && self.hash_type == HashType::Ed25519
  }

  /// Check whether the header is formatted as a `.tree`.
  pub fn is_tree(&self) -> bool {
    self.entry_size == 40
      && self.file_type == FileType::Tree
      && self.hash_type == HashType::BLAKE2b
  }
}


src/parsers.rs
#![cfg_attr(feature = "cargo-clippy", allow(clippy))]

use header::*;
use nom::{be_u16, be_u8, rest};
use std::str;

pub(crate) const HEADER_LENGTH: usize = 32;
const VERIFY_TRAILING_ZEROS: bool = true;

named!(
  file_type<FileType>,
  switch!(be_u8,
    0 => value!(FileType::BitField) |
    1 => value!(FileType::Signatures) |
    2 => value!(FileType::Tree)
  )
);

named!(
  protocol_version<ProtocolVersion>,
  switch!(be_u8,
    0 => value!(ProtocolVersion::V0)
  )
);

named_args!(
  algorithm(len: u8)<HashType>,
  switch!(map_res!(take!(len), str::from_utf8),
    "BLAKE2b" => value!(HashType::BLAKE2b) |
    "Ed25519" => value!(HashType::Ed25519) |
    "" => value!(HashType::None)
  )
);

named!(
  pub header<Header>,
  flat_map!(
    take!(HEADER_LENGTH),
    do_parse!(
      tag!(b"\x05\x02\x57") >>
      file_type: file_type >>
      protocol_version: protocol_version >>
      entry_size: be_u16 >>

      algorithm_len: verify!(be_u8, |len: u8| len <= HEADER_LENGTH as u8 - 8) >>
      algorithm: apply!(algorithm, algorithm_len) >>

      verify!(rest, |bytes: &[u8]| {
        let header_consumed = bytes.len() + algorithm_len as usize + 8 == HEADER_LENGTH;
        let trailing_zeros = !VERIFY_TRAILING_ZEROS || bytes.iter().all(|&b| b == 0u8);
        header_consumed && trailing_zeros
      }) >>

      (Header {
        file_type,
        protocol_version,
        entry_size,
        hash_type: algorithm,
      })
    )
  )
);

#[cfg(test)]
mod test {
  use super::*;

  use nom;

  #[test]
  fn parse_file_type() {
    assert_eq!(
      file_type(b"\x00"),
      Ok((&[][..], FileType::BitField))
    );
    assert_eq!(
      file_type(b"\x01"),
      Ok((&[][..], FileType::Signatures))
    );
    assert_eq!(
      file_type(b"\x02"),
      Ok((&[][..], FileType::Tree))
    );
    assert!(file_type(b"\xff").is_err());
  }

  #[test]
  fn parse_header() {
    fn mk_header(prefix: &[u8]) -> [u8; 32] {
      let mut h = [0u8; 32];
      h[0..prefix.len()].clone_from_slice(prefix);
      h
    }

    assert_eq!(
      header(&mk_header(
        b"\x05\x02W\x01\x00\x00\x28\x07BLAKE2b"
      )),
      Ok((
        &[][..],
        Header {
          file_type: FileType::Signatures,
          protocol_version: ProtocolVersion::V0,
          entry_size: 40,
          hash_type: HashType::BLAKE2b
        }
      ))
    );
    assert_eq!(
      header(&mk_header(
        b"\x05\x02W\x01\x00\x00\x28\x07BLAKE2b"
      )).unwrap()
        .1
        .hash_type,
      HashType::BLAKE2b
    );
    assert_eq!(
      header(&mk_header(
        b"\x05\x02W\x01\x00\x00\x28\x07Ed25519"
      )).unwrap()
        .1
        .hash_type,
      HashType::Ed25519
    );
    assert_eq!(
      header(&mk_header(b"\x05\x02W\x01\x00\x00\x28\x00"))
        .unwrap()
        .1
        .hash_type,
      HashType::None
    );
    assert!(header(&mk_header(b"\x05\x02W\x01\x00\x00\x28\x01B")).is_err());
    assert!(header(&mk_header(b"\x05\x02W\x01\x00\x00\x28\x01B")).is_err());

    let h = b"\x05\x02W\x01\x00\x00\x28\x19BLAKE2bXXXXXXXXXXXXXXXXXX";
    assert!(header(h).is_err());
  }

  #[test]
  fn invalid_algorithm_len() {
    match header(b"\x05\x02W\x00\x00\x00\x00\xFF\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00") {
      Err(nom::Err::Error(nom::Context::Code(_, nom::ErrorKind::Verify))) => (),
      x => panic!("{:?}", x),
    }
  }
}


src/file_type.rs
/// Type of file.
///
/// `signatures`, `bitfield` and `tree` are the three SLEEP files. There are two
///additional files, `key`, and `data`, which do not contain SLEEP file headers
///and store plain serialized data for easy access. `key` stores the public key
///that is described by the `signatures` file, and `data` stores the raw chunk
///data that the `tree` file contains the hashes and metadata.
#[derive(Debug, PartialEq)]
pub enum FileType {
  /// The bitfield describes which pieces of data you have, and which nodes in
  /// the tree file have been written.  This file exists as an index of the tree
  /// and data to quickly figure out which pieces of data you have or are
  /// missing. This file can be regenerated if you delete it, so it is
  /// considered a materialized index.
  BitField,
  /// A SLEEP formatted 32 byte header with data entries being 64 byte
  /// signatures.
  Signatures,
  /// A SLEEP formatted 32 byte header with data entries representing a
  /// serialized Merkle tree based on the data in the data storage layer. All
  /// the fixed size nodes written in in-order tree notation. The header
  /// algorithm string for `tree` files is `BLAKE2b`. The entry size is 40
  /// bytes.
  Tree,
}

impl FileType {
  /// Returns true if the file is `BitField`
  #[inline]
  pub fn is_bitfield(&self) -> bool {
    *self == FileType::BitField
  }

  /// Returns true if the file is `Signatures`
  #[inline]
  pub fn is_signatures(&self) -> bool {
    *self == FileType::Signatures
  }

  /// Returns true if the file is `Tree`
  #[inline]
  pub fn is_tree(&self) -> bool {
    *self == FileType::Tree
  }
}


src/lib.rs
#![forbid(unsafe_code, missing_debug_implementations, missing_docs)]
#![cfg_attr(test, deny(warnings))]

//! ## Example
//! ```rust
//! extern crate sleep_parser as sleep_parser;
//!
//! use sleep_parser::{FileType, HashType, Header};
//! use std::fs::File;
//! use std::io::{BufRead, BufReader};
//!
//! let file = File::open("tests/fixtures/content.bitfield").unwrap();
//! let mut reader = BufReader::with_capacity(32, file);
//! let buffer = reader.fill_buf().unwrap();
//! let header = Header::from_vec(&buffer).unwrap();
//! assert!(header.is_bitfield());
//! ```

mod file_type;
mod hash_type;
mod protocol_version;

use byteorder::{BigEndian, ReadBytesExt, WriteBytesExt};
use failure::{bail, ensure, format_err, Error};
use std::io::Cursor;

pub use crate::file_type::FileType;
pub use crate::hash_type::HashType;
pub use crate::protocol_version::ProtocolVersion;

const HEADER_LENGTH: usize = 32;
const MAX_ALGORITHM_NAME_LENGTH: usize = HEADER_LENGTH - 8;

/// Structural representation of 32 byte SLEEP headers.
#[derive(Debug)]
pub struct Header {
  /// Type of file.
  pub file_type: FileType,
  /// Version of the SLEEP protocol.
  pub protocol_version: ProtocolVersion,
  /// Size of each piece of data in the file body.
  pub entry_size: u16,
  /// Algorithm used for hashing the content.
  pub hash_type: HashType,
}

impl Header {
  /// Create a new `Header`.
  #[inline]
  pub fn new(
    file_type: FileType,
    entry_size: u16,
    hash_type: HashType,
  ) -> Self {
    Header {
      file_type,
      entry_size,
      hash_type,
      protocol_version: ProtocolVersion::V0,
    }
  }

  /// Parse a 32 byte buffer slice into a valid Header.
  #[inline]
  pub fn from_vec(buffer: &[u8]) -> Result<Header, Error> {
    ensure!(buffer.len() == 32, "buffer should be 32 bytes");

    let mut rdr = Cursor::new(buffer);
    let byte = rdr.read_u8().unwrap();
    ensure!(
      byte == 5,
      "The first byte of a SLEEP header should be '5', found {}",
      byte
    );

    let byte = rdr.read_u8().unwrap();
    ensure!(
      byte == 2,
      "The second byte of a SLEEP header should be '2', found {}",
      byte
    );

    let byte = rdr.read_u8().unwrap();
    ensure!(
      byte == 87,
      "The third byte of a SLEEP header should be '87', found {}",
      byte
    );

    let file_type = match rdr.read_u8().unwrap() {
      0 => FileType::BitField,
      1 => FileType::Signatures,
      2 => FileType::Tree,
      num => bail!(
        "The fourth byte '{}' does not belong to any known SLEEP file type",
        num
      ),
    };

    let protocol_version = match rdr.read_u8().unwrap() {
      0 => ProtocolVersion::V0,
      num => bail!(
        "The fifth byte '{}' does not belong to any known SLEEP protocol protocol_version",
        num
      ),
    };

    // Read entry size which will inform how many bytes to read next.
    let entry_size = rdr.read_u16::<BigEndian>().unwrap();

    // Read out the "entry_size" bytes into a string.
    // NOTE(yw): there should be a more concise way of doing this.
    let hash_name_len = rdr.read_u8().unwrap() as usize;
    let current = rdr.position() as usize;

    ensure!(
      hash_name_len <= MAX_ALGORITHM_NAME_LENGTH,
      "Algorithm name is too long: {} (max: {})",
      hash_name_len,
      MAX_ALGORITHM_NAME_LENGTH
    );

    let hash_name_upper = current + hash_name_len;
    ensure!(
      buffer.len() >= hash_name_upper,
      "Broken parser: algorithm name is out of bounds: {} {}",
      hash_name_upper,
      buffer.len()
    );

    let buf_slice = &buffer[current..hash_name_upper];
    rdr.set_position(hash_name_upper as u64 + 1);
    let algo = ::std::str::from_utf8(buf_slice).map_err(|e| {
      format_err!("The algorithm string was invalid utf8 encoded: {:?}", e)
    })?;

    let hash_type = match algo {
      "BLAKE2b" => HashType::BLAKE2b,
      "Ed25519" => HashType::Ed25519,
      "" => HashType::None,
      name => bail!("Unexpected algorithm name: {}", name),
    };

    Ok(Header {
      protocol_version,
      entry_size,
      file_type,
      hash_type,
    })
  }

  /// Convert a `Header` into a `Vec<u8>`. Use this to persist a header back to
  /// disk.
  #[inline]
  pub fn to_vec(&self) -> Vec<u8> {
    let mut wtr = Vec::with_capacity(32);

    wtr.extend_from_slice(&[5u8, 2, 87]);

    let file_type = match self.file_type {
      FileType::BitField => 0,
      FileType::Signatures => 1,
      FileType::Tree => 2,
    };
    wtr.write_u8(file_type).unwrap();

    let protocol_version = match self.protocol_version {
      ProtocolVersion::V0 => 0,
    };
    wtr.write_u8(protocol_version).unwrap();

    wtr.write_u16::<BigEndian>(self.entry_size).unwrap();

    let hash_type = match self.hash_type {
      HashType::BLAKE2b => "BLAKE2b",
      HashType::Ed25519 => "Ed25519",
      HashType::None => "",
    };
    let hash_type = hash_type.as_bytes();
    wtr.write_u8(hash_type.len() as u8).unwrap();
    wtr.extend_from_slice(hash_type);

    for _ in wtr.len()..wtr.capacity() {
      wtr.write_u8(0).unwrap();
    }
    wtr
  }

  /// Check whether the header is formatted as a `.bitfield`.
  #[inline]
  pub fn is_bitfield(&self) -> bool {
    self.entry_size == 3328
      && self.file_type.is_bitfield()
      && self.hash_type.is_none()
  }

  /// Check whether the header is formatted as a `.signatures`.
  #[inline]
  pub fn is_signatures(&self) -> bool {
    self.entry_size == 64
      && self.file_type.is_signatures()
      && self.hash_type.is_ed25519()
  }

  /// Check whether the header is formatted as a `.tree`.
  #[inline]
  pub fn is_tree(&self) -> bool {
    self.entry_size == 40
      && self.file_type.is_tree()
      && self.hash_type.is_blake2b()
  }
}

/// Create a new `Header` in the `Bitfield` configuration.
#[inline]
pub fn create_bitfield() -> Header {
  Header::new(FileType::BitField, 3328, HashType::None)
}

/// Create a new `Header` in the `Signatures` configuration.
#[inline]
pub fn create_signatures() -> Header {
  Header::new(FileType::Signatures, 64, HashType::Ed25519)
}

/// Create a new `Header` in the `Tree` configuration.
#[inline]
pub fn create_tree() -> Header {
  Header::new(FileType::Tree, 40, HashType::BLAKE2b)
}
